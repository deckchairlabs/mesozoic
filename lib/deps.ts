export {
  copy,
  emptyDir,
  ensureDir,
  walk,
} from "https://deno.land/std@0.153.0/fs/mod.ts";
export {
  cache,
  exists as existsInCache,
} from "https://deno.land/x/cache@0.2.13/mod.ts";
export { crayon } from "https://deno.land/x/crayon@3.3.2/mod.ts";
export * as log from "https://deno.land/std@0.153.0/log/mod.ts";
export {
  createGraph,
  load as graphDefaultLoad,
} from "https://deno.land/x/deno_graph@0.32.0/mod.ts";
export { deepMerge } from "https://deno.land/std@0.153.0/collections/deep_merge.ts";
export {
  init as initModuleLexer,
  parse as parseModule,
} from "https://esm.sh/es-module-lexer@1.0.3/es-module-lexer.js";
export type { ImportSpecifier } from "https://esm.sh/v95/es-module-lexer@1.0.3/types/lexer.d.ts";
export {
  parse as parseImportMap,
  resolve as importMapResolve,
} from "https://esm.sh/@import-maps/resolve@1.0.1/resolve.js";

export { sprintf } from "https://deno.land/std@0.153.0/fmt/printf.ts";

export {
  basename,
  dirname,
  extname,
  fromFileUrl,
  globToRegExp,
  isAbsolute,
  join,
  normalize,
  relative,
  resolve,
  SEP,
  toFileUrl,
} from "https://deno.land/std@0.153.0/path/mod.ts";
