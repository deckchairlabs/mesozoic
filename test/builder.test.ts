import { log } from "../lib/deps.ts";
import { BuildContext, Builder, ContextBuilder } from "../mod.ts";
import { assertEquals, assertSnapshot } from "./deps.ts";
import { getFixtureDir, getOutputDir } from "./helpers.ts";

const outputDir = getOutputDir("app");

async function createBuilder(context: BuildContext) {
  const logLevel = Deno.env.get("MESOZOIC_LOG");
  const builder = new Builder(context, {
    logLevel: logLevel as log.LevelName,
  });

  await builder.cleanOutput();

  return builder;
}

Deno.test(
  "it can copy, compile and vendor entrypoints producing valid import maps",
  async (t) => {
    const context = new ContextBuilder()
      .setRoot(getFixtureDir("app"))
      .setOutput(outputDir)
      .setImportMapPath("./importMap.json")
      .ignore([
        "./README.md",
        "./.private/**/*",
        "./.git/**/*",
      ])
      .contentHash([
        "./**/*.+(ts|tsx|js|jsx|css|jpg)",
        "!./vendor/server/**/*",
        "!./server.+(ts|tsx|js|jsx)",
      ])
      .compile([
        "./**/*.+(ts|tsx|js|jsx)",
        "!./vendor/server/**/*",
      ])
      .build();

    const builder = await createBuilder(context);

    builder.setEntrypoints({
      "browser": {
        path: "./client.tsx",
        target: "browser",
      },
      "server": {
        path: "./server.tsx",
        target: "deno",
      },
    });

    const sources = await builder.gatherSources();
    const result = await builder.build(sources);

    const vendored = result.outputSources.filter((source) =>
      source.relativePath().startsWith("./vendor")
    );

    assertEquals(result.outputSources.size > 0, true);
    assertEquals(vendored.size > 0, true);
    assertEquals(result.importMaps.size, 2);

    assertEquals(
      builder.toManifest(result.outputSources, {
        prefix: "/",
        ignore: ["./**/*", "!./public/**/*"],
      }).length,
      3,
    );

    for (const [, importMap] of result.importMaps) {
      await assertSnapshot(t, importMap);
    }
  },
);

Deno.test("it can copy and compile entrypoints producing valid import maps", async (t) => {
  const context = new ContextBuilder()
    .setRoot(getFixtureDir("app"))
    .setOutput(outputDir)
    .setImportMapPath("./importMap.json")
    .setVendorDependencies(false)
    .ignore([
      "./README.md",
      "./.private/**/*",
      "./.git/**/*",
    ])
    .contentHash([
      "./**/*.+(ts|tsx|js|jsx|css|jpg|woff2|woff|ttf|svg|png)",
      "!./vendor/server/**/*",
      "!./server.+(ts|tsx|js|jsx)",
    ])
    .compile([
      "./**/*.+(ts|tsx|js|jsx)",
      "!./vendor/server/**/*",
    ])
    .build();

  const builder = await createBuilder(context);

  builder.setEntrypoints({
    "browser": {
      path: "./client.tsx",
      target: "browser",
    },
    "server": {
      path: "./server.tsx",
      target: "deno",
    },
  });

  const sources = await builder.gatherSources();
  const result = await builder.build(sources);

  const vendored = result.outputSources.filter((source) =>
    source.relativePath().startsWith("./vendor")
  );

  assertEquals(result.outputSources.size > 0, true);
  assertEquals(vendored.size === 0, true);
  assertEquals(result.importMaps.size, 2);

  // const stylesheetSource = result.outputSources.find((source) =>
  //   source.originalPath().relativePath() === "./public/styles/main.css"
  // );

  // assert(stylesheetSource);
  // console.log(stylesheetSource);

  // const stylesheet = await stylesheetSource.read();
  // assertEquals(stylesheet.includes('@import "components.'), true);
  // assertEquals(stylesheet.includes('url("../font.'), true);
  // assertEquals(stylesheet.includes('url("../image.'), true);

  assertEquals(
    builder.toManifest(result.outputSources, {
      prefix: "/",
      ignore: ["./**/*", "!./public/**/*"],
    }).length,
    4,
  );

  for (const [, importMap] of result.importMaps) {
    await assertSnapshot(t, importMap);
  }
});
