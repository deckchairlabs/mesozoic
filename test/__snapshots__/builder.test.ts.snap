export const snapshot = {};

snapshot[`it can copy, compile and vendor entrypoints producing valid import maps 1`] = `
{
  imports: {
    "./client.tsx": "/client.d18310e0.js",
    "./src/app.tsx": "/src/app.d82ab43c.js",
    "./src/components/Test.tsx": "/src/components/Test.4e662968.js",
    "https://deno.land/": "/vendor/browser/deno.land/",
    "https://esm.sh/": "/vendor/browser/esm.sh/",
    react: "/vendor/browser/esm.sh/stable/react@18.2.0/es2022/react.625c3f59.js",
    "react-dom/client": "/vendor/browser/esm.sh/v93/react-dom@18.2.0/es2022/client.2652ccc5.js",
    "react/jsx-runtime": "/vendor/browser/esm.sh/stable/react@18.2.0/es2022/jsx-runtime.9bacc56c.js",
    twind: "/vendor/browser/esm.sh/v103/twind@0.16.17/es2022/twind.a56db6c9.js",
    "ultra/hooks/use-asset.js": "/vendor/browser/deno.land/x/ultra@v2.1.7/hooks/use-asset.d43ee186.js",
    "ultra/hydrate.js": "/vendor/browser/deno.land/x/ultra@v2.1.7/hydrate.c1c2367e.js",
  },
  scopes: {
    "/vendor/browser/deno.land/": {
      "/x/": "/vendor/browser/deno.land/x/",
    },
    "/vendor/browser/esm.sh/": {
      "/stable/": "/vendor/browser/esm.sh/stable/",
      "/v103/": "/vendor/browser/esm.sh/v103/",
      "/v93/": "/vendor/browser/esm.sh/v93/",
    },
  },
}
`;

snapshot[`it can copy, compile and vendor entrypoints producing valid import maps 2`] = `
{
  imports: {
    "./server.tsx": "./server.js",
    "./src/app.tsx": "./src/app.d82ab43c.js",
    "./src/components/Test.tsx": "./src/components/Test.4e662968.js",
    "https://deno.land/": "./vendor/server/deno.land/",
    "https://esm.sh/": "./vendor/server/esm.sh/",
    react: "./vendor/server/esm.sh/stable/react@18.2.0/deno/react.js",
    "react-dom/server": "./vendor/server/esm.sh/v93/react-dom@18.2.0/deno/server.js",
    "react/jsx-runtime": "./vendor/server/esm.sh/stable/react@18.2.0/deno/jsx-runtime.js",
    twind: "./vendor/server/esm.sh/v103/twind@0.16.17/deno/twind.js",
    "ultra/hooks/use-asset.js": "./vendor/server/deno.land/x/ultra@v2.1.7/hooks/use-asset.js",
    "ultra/server.ts": "./vendor/server/deno.land/x/ultra@v2.1.7/server.ts",
  },
  scopes: {
    "./vendor/server/deno.land/": {
      "/std@0.163.0/": "./vendor/server/deno.land/std@0.163.0/",
      "/std@0.164.0/": "./vendor/server/deno.land/std@0.164.0/",
      "/std@0.173.0/": "./vendor/server/deno.land/std@0.173.0/",
      "/std@0.97.0/": "./vendor/server/deno.land/std@0.97.0/",
      "/x/": "./vendor/server/deno.land/x/",
    },
    "./vendor/server/esm.sh/": {
      "/stable/": "./vendor/server/esm.sh/stable/",
      "/v103/": "./vendor/server/esm.sh/v103/",
      "/v93/": "./vendor/server/esm.sh/v93/",
    },
  },
}
`;

snapshot[`it can copy and compile entrypoints producing valid import maps 1`] = `
{
  imports: {
    "./client.tsx": "/client.d18310e0.js",
    "./src/app.tsx": "/src/app.d82ab43c.js",
    "./src/components/Test.tsx": "/src/components/Test.4e662968.js",
    "https://deno.land/": "https://deno.land/",
    "https://esm.sh/": "https://esm.sh/",
    react: "https://esm.sh/stable/react@18.2.0/es2022/react.js",
    "react-dom/client": "https://esm.sh/v93/react-dom@18.2.0/es2022/client.js",
    "react/jsx-runtime": "https://esm.sh/stable/react@18.2.0/es2022/jsx-runtime.js",
    twind: "https://esm.sh/v103/twind@0.16.17/es2022/twind.js",
    "ultra/hooks/use-asset.js": "https://deno.land/x/ultra@v2.1.7/hooks/use-asset.js",
    "ultra/hydrate.js": "https://deno.land/x/ultra@v2.1.7/hydrate.js",
  },
  scopes: {
    "https://deno.land/": {
      "/x/": "https://deno.land/x/",
    },
    "https://esm.sh/": {
      "/stable/": "https://esm.sh/stable/",
      "/v103/": "https://esm.sh/v103/",
      "/v93/": "https://esm.sh/v93/",
    },
  },
}
`;

snapshot[`it can copy and compile entrypoints producing valid import maps 2`] = `
{
  imports: {
    "./server.tsx": "./server.js",
    "./src/app.tsx": "./src/app.d82ab43c.js",
    "./src/components/Test.tsx": "./src/components/Test.4e662968.js",
    "https://deno.land/": "https://deno.land/",
    "https://esm.sh/": "https://esm.sh/",
    react: "https://esm.sh/stable/react@18.2.0/deno/react.js",
    "react-dom/server": "https://esm.sh/v93/react-dom@18.2.0/deno/server.js",
    "react/jsx-runtime": "https://esm.sh/stable/react@18.2.0/deno/jsx-runtime.js",
    twind: "https://esm.sh/v103/twind@0.16.17/deno/twind.js",
    "ultra/hooks/use-asset.js": "https://deno.land/x/ultra@v2.1.7/hooks/use-asset.js",
    "ultra/server.ts": "https://deno.land/x/ultra@v2.1.7/server.ts",
  },
  scopes: {
    "https://deno.land/": {
      "/std@0.163.0/": "https://deno.land/std@0.163.0/",
      "/std@0.164.0/": "https://deno.land/std@0.164.0/",
      "/std@0.173.0/": "https://deno.land/std@0.173.0/",
      "/std@0.97.0/": "https://deno.land/std@0.97.0/",
      "/x/": "https://deno.land/x/",
    },
    "https://esm.sh/": {
      "/stable/": "https://esm.sh/stable/",
      "/v103/": "https://esm.sh/v103/",
      "/v93/": "https://esm.sh/v93/",
    },
  },
}
`;
